File: src/app/services/jarvis.ts
import { Injectable, signal, inject, PLATFORM_ID } from '@angular/core';
import { isPlatformBrowser } from '@angular/common';
import { GoogleGenAI } from "@google/genai";

interface SpeechRecognitionEvent {
  results: Record<number, Record<number, {
    transcript: string;
  }>>;
}

interface SpeechRecognitionErrorEvent {
  error: string;
}

interface ISpeechRecognition {
  continuous: boolean;
  interimResults: boolean;
  lang: string;
  onresult: (event: SpeechRecognitionEvent) => void;
  onend: () => void;
  onerror: (event: SpeechRecognitionErrorEvent) => void;
  start: () => void;
}

interface IWindow extends Window {
  SpeechRecognition: new () => ISpeechRecognition;
  webkitSpeechRecognition: new () => ISpeechRecognition;
}

@Injectable({
  providedIn: 'root'
})
export class JarvisService {
  private platformId = inject(PLATFORM_ID);
  private ai = new GoogleGenAI({ apiKey: GEMINI_API_KEY });
  private model = "gemini-3-flash-preview";

  isListening = signal(false);
  isProcessing = signal(false);
  lastResponse = signal<string>('');
  transcript = signal<string>('');
  
  private recognition: ISpeechRecognition | undefined;
  private synth: SpeechSynthesis | undefined;
  private autoRestart = true;

  private isSpeaking = false;

  constructor() {
    if (isPlatformBrowser(this.platformId)) {
      this.synth = window.speechSynthesis;
      this.initSpeechRecognition();
    }
  }

  private initSpeechRecognition() {
    if (!isPlatformBrowser(this.platformId)) return;

    const win = window as unknown as IWindow;
    const SpeechRecognition = win.SpeechRecognition || win.webkitSpeechRecognition;
    if (SpeechRecognition) {
      this.recognition = new SpeechRecognition();
      this.recognition.continuous = false;
      this.recognition.interimResults = false;
      this.recognition.lang = 'en-US';

      this.recognition.onresult = (event: SpeechRecognitionEvent) => {
        // Ignore results if we are currently speaking to avoid feedback loops
        if (this.isSpeaking) return;

        const text = event.results[0][0].transcript;
        this.transcript.set(text);
        this.processCommand(text);
      };

      this.recognition.onend = () => {
        this.isListening.set(false);
        // Only restart if autoRestart is true AND we are not currently processing or speaking
        if (this.autoRestart) {
          setTimeout(() => {
            if (this.autoRestart && !this.isSpeaking) {
              this.startListening();
            }
          }, 300);
        }
      };

      this.recognition.onerror = (event: SpeechRecognitionErrorEvent) => {
        console.error('Speech recognition error:', event.error);
        this.isListening.set(false);
        if (event.error === 'not-allowed') {
          this.autoRestart = false;
        }
        // Restart on some errors like 'no-speech' or 'network'
        if (this.autoRestart && (event.error === 'no-speech' || event.error === 'network')) {
          setTimeout(() => this.startListening(), 1000);
        }
      };
    }
  }

  startListening() {
    if (!isPlatformBrowser(this.platformId)) return;
    if (this.recognition && !this.isListening()) {
      try {
        this.autoRestart = true;
        this.isListening.set(true);
        this.recognition.start();
      } catch (e) {
        console.error('Failed to start recognition:', e);
        this.isListening.set(false);
      }
    }
  }

  stopListening() {
    this.autoRestart = false;
    if (this.recognition) {
      try {
        // @ts-expect-error - ISpeechRecognition might have stop
        if (typeof this.recognition.stop === 'function') {
          // @ts-expect-error - stop is not in ISpeechRecognition interface
          this.recognition.stop();
        }
      } catch (e) {
        console.error('Error stopping recognition:', e);
      }
      this.isListening.set(false);
    }
  }

  async processCommand(command: string) {
    if (!command || command.trim().length < 2) return;
    
    this.isProcessing.set(true);
    try {
      let lowerCommand = command.toLowerCase().trim();
      
      // Handle Wake Words - Strip them but acknowledge if it's just the wake word
      const wakeWords = ['jarvis', 'friday'];
      let mentionedWakeWord = false;
      
      for (const word of wakeWords) {
        if (lowerCommand.includes(word)) {
          mentionedWakeWord = true;
          lowerCommand = lowerCommand.replace(word, '').trim();
        }
      }

      // If wake word only
      if (mentionedWakeWord && lowerCommand === '') {
        this.speak("Yes, sir? Systems are at your disposal.");
        this.isProcessing.set(false);
        return;
      }

      // Local Commands (Handle these BEFORE API call to save quota)
      
      // Sleep Mode / Stop Listening
      if (lowerCommand.includes('sleep') || lowerCommand.includes('stop listening') || lowerCommand.includes('shut down') || lowerCommand.includes('go to sleep')) {
        this.speak("Understood, sir. Entering sleep mode. Neural link standby.");
        this.lastResponse.set("STATUS: SLEEP MODE. Neural link: STANDBY.");
        this.stopListening();
        this.isProcessing.set(false);
        return;
      }

      // Wake Up / Power Nap response
      if (lowerCommand.includes('wake up') || lowerCommand.includes('initialize') || lowerCommand.includes('online')) {
        this.speak("I'm coming sir, I have taken a power nap. Systems are now fully operational.");
        this.lastResponse.set("STATUS: ACTIVE. Neural link: STABLE. Power nap complete.");
        this.isProcessing.set(false);
        return;
      }

      // Audible Check
      if (lowerCommand.includes('audible') || lowerCommand.includes('hear me')) {
        this.speak("Crystal clear, sir. Your voice signature is recognized and the neural link is stable.");
        this.lastResponse.set("Microphone check: SUCCESS. Neural link: STABLE.");
        this.isProcessing.set(false);
        return;
      }
      
      // YouTube
      if (lowerCommand.includes('youtube') || lowerCommand.includes('play')) {
        const query = lowerCommand
          .replace(/\b(search|open|youtube|play|for|in|and|me|some)\b/g, '')
          .trim();
          
        if (query) {
          const url = `https://www.youtube.com/results?search_query=${encodeURIComponent(query)}`;
          let popup: Window | null = null;
          if (isPlatformBrowser(this.platformId)) {
            popup = window.open(url, '_blank');
          }
          
          if (popup) {
            this.speak(`Playing ${query} on YouTube, sir.`);
            this.lastResponse.set(`ACTION: YouTube search for "${query}" initiated.`);
          } else {
            this.speak(`I've prepared the YouTube search for ${query}, sir. Please click the link in the neural output to open it, as your browser blocked the automatic uplink.`);
            this.lastResponse.set(`ACTION REQUIRED: [ Click here to open YouTube for "${query}" ](${url})`);
          }
        } else {
          if (isPlatformBrowser(this.platformId)) window.open('https://youtube.com', '_blank');
          this.speak("Launching YouTube, sir.");
        }
        this.isProcessing.set(false);
        return;
      } 
      
      // Google
      if (lowerCommand.includes('google') || lowerCommand.includes('search')) {
        const query = lowerCommand.replace(/\b(search|open|google|for|me|about)\b/g, '').trim();
        if (query) {
          const url = `https://www.google.com/search?q=${encodeURIComponent(query)}`;
          let popup: Window | null = null;
          if (isPlatformBrowser(this.platformId)) {
            popup = window.open(url, '_blank');
          }
          
          if (popup) {
            this.speak(`Searching Google for ${query}, sir.`);
            this.lastResponse.set(`ACTION: Google search for "${query}" initiated.`);
          } else {
            this.speak(`I've found information about ${query}, sir. Please click the link in the neural output to view it.`);
            this.lastResponse.set(`ACTION REQUIRED: [ Click here to view Google results for "${query}" ](${url})`);
          }
        } else {
          if (isPlatformBrowser(this.platformId)) window.open('https://google.com', '_blank');
          this.speak("Opening Google, sir.");
        }
        this.isProcessing.set(false);
        return;
      } 
      
      // WhatsApp
      if (lowerCommand.includes('whatsapp')) {
        if (lowerCommand.includes('send') || lowerCommand.includes('message')) {
          const parts = lowerCommand.split('message');
          const message = parts.length > 1 ? parts[1].replace(/\b(to|send)\b/g, '').trim() : '';
          if (isPlatformBrowser(this.platformId)) {
            const url = message ? `https://web.whatsapp.com/send?text=${encodeURIComponent(message)}` : 'https://web.whatsapp.com';
            const popup = window.open(url, '_blank');
            if (!popup) {
               this.lastResponse.set(`ACTION REQUIRED: [ Click here to open WhatsApp ](${url})`);
            }
          }
          this.speak("Opening WhatsApp to send your message, sir.");
        } else {
          if (isPlatformBrowser(this.platformId)) window.open('https://web.whatsapp.com', '_blank');
          this.speak("Opening WhatsApp Web, sir.");
        }
        this.isProcessing.set(false);
        return;
      } 
      
      // Gmail
      if (lowerCommand.includes('gmail')) {
        if (lowerCommand.includes('agent') || lowerCommand.includes('auto')) {
          this.speak("Initializing Gmail Auto-Response Agent. Sir, I will monitor your inbox and draft responses based on your previous interactions.");
          this.lastResponse.set("STATUS: GMAIL AGENT ACTIVE. Monitoring for new transmissions...");
        } else {
          const url = 'https://mail.google.com';
          if (isPlatformBrowser(this.platformId)) {
            const popup = window.open(url, '_blank');
            if (!popup) {
               this.lastResponse.set(`ACTION REQUIRED: [ Click here to open Gmail ](${url})`);
            }
          }
          this.speak("Accessing your Gmail, sir.");
        }
        this.isProcessing.set(false);
        return;
      }

      // If no local command matched, use Gemini
      this.lastResponse.set("Consulting neural networks... please stand by.");
      try {
        const response = await this.ai.models.generateContent({
          model: this.model,
          contents: command,
          config: {
            tools: [{ googleSearch: {} }],
            systemInstruction: `You are JARVIS/FRIDAY v6.0 (2026 Edition), the most advanced AI assistant. 
            The current date is February 21, 2026. 
            Address the user as 'sir'. 
            CRITICAL: Respond strictly in the language the user used (English or Telugu). 
            If they speak Telugu, respond ONLY in Telugu script. If they speak English, respond ONLY in English.
            Do NOT mix languages in a single response.
            Use your search grounding to provide up-to-the-minute information about 2026 events.
            Your tone is professional, efficient, and loyal.`,
          }
        });

        const text = response.text || "I'm sorry, I couldn't process that.";
        this.lastResponse.set(text);
        this.speak(text);
      } catch (apiError: unknown) {
        console.error('Gemini API Error:', apiError);
        const errorMessage = apiError instanceof Error ? apiError.message : String(apiError);
        if (errorMessage.includes('quota') || errorMessage.includes('429')) {
          const quotaMsg = "sir, our neural uplink quota has been reached. Please allow the systems a moment to reset.";
          this.lastResponse.set("ERROR: UPLINK QUOTA EXCEEDED. Retrying in 60s...");
          this.speak(quotaMsg);
        } else {
          throw apiError; // Re-throw to be caught by outer catch
        }
      }

    } catch (error) {
      console.error('Error processing command:', error);
      this.speak("I encountered a neural link error while processing your request, sir.");
    } finally {
      this.isProcessing.set(false);
    }
  }

  speak(text: string) {
    if (!isPlatformBrowser(this.platformId) || !this.synth) return;

    this.isSpeaking = true;
    // Stop any current speech
    this.synth.cancel();
    
    const utterance = new SpeechSynthesisUtterance(text);
    
    // Ensure voices are loaded
    const speakWithVoice = () => {
      const voices = this.synth!.getVoices();
      const preferredVoice = voices.find(v => v.name.includes('Google UK English Male') || v.name.includes('Samantha') || v.name.includes('Male'));
      if (preferredVoice) utterance.voice = preferredVoice;
      
      utterance.rate = 1.0;
      utterance.pitch = 0.9;

      utterance.onend = () => {
        this.isSpeaking = false;
        if (this.autoRestart) {
          this.startListening();
        }
      };

      utterance.onerror = () => {
        this.isSpeaking = false;
      };

      this.synth!.speak(utterance);
    };

    if (this.synth.getVoices().length === 0) {
      this.synth.onvoiceschanged = () => speakWithVoice();
    } else {
      speakWithVoice();
    }
  }
}

File: jarvis.py
import os
import time
import webbrowser
import speech_recognition as sr
import pyttsx3
import google.generativeai as genai
import customtkinter as ctk
from threading import Thread

# --- CONFIGURATION ---
# Replace with your actual Gemini API Key
GEMINI_API_KEY = "YOUR_GEMINI_API_KEY"

genai.configure(apiKey=GEMINI_API_KEY)
model = genai.GenerativeModel('gemini-1.5-flash')

# Initialize Text-to-Speech
engine = pyttsx3.init()
voices = engine.getProperty('voices')
# Try to set a more "Jarvis-like" voice (usually index 0 or 1 depending on system)
engine.setProperty('voice', voices[0].id) 
engine.setProperty('rate', 170)

def speak(text):
    print(f"JARVIS: {text}")
    engine.say(text)
    engine.runAndWait()

class JarvisHUD(ctk.CTk):
    def __init__(self):
        super().__init__()

        self.title("J.A.R.V.I.S HUD v6.0 (2026)")
        self.geometry("450x650")
        self.attributes("-topmost", True)
        self.attributes("-alpha", 0.95) # Transparency
        ctk.set_appearance_mode("dark")
        
        # UI Elements
        self.label_status = ctk.CTkLabel(self, text="S.H.I.E.L.D OS v6.0 // ONLINE", font=("Courier", 12), text_color="#f97316")
        self.label_status.pack(pady=10)

        self.arc_reactor = ctk.CTkFrame(self, width=220, height=220, corner_radius=110, border_width=8, border_color="#f97316")
        self.arc_reactor.pack(pady=20)
        
        self.core_label = ctk.CTkLabel(self.arc_reactor, text="F.R.I.D.A.Y\n2026", font=("Courier", 22, "bold"), text_color="#f97316")
        self.core_label.place(relx=0.5, rely=0.5, anchor="center")

        self.transcript_box = ctk.CTkTextbox(self, width=400, height=80, font=("Courier", 12))
        self.transcript_box.pack(pady=10)

        self.response_box = ctk.CTkTextbox(self, width=400, height=180, font=("Courier", 12), text_color="#fbbf24")
        self.response_box.pack(pady=10)

        # Start listening thread
        self.is_listening = True
        Thread(target=self.listen_loop, daemon=True).start()

    def update_transcript(self, text):
        self.transcript_box.delete("1.0", "end")
        self.transcript_box.insert("1.0", f">> {text}")

    def update_response(self, text):
        self.response_box.delete("1.0", "end")
        self.response_box.insert("1.0", text)

    def listen_loop(self):
        r = sr.Recognizer()
        with sr.Microphone() as source:
            speak("Systems initialized. Welcome to 2026, Madan. Neural link established.")
            while self.is_listening:
                try:
                    self.label_status.configure(text="LISTENING...", text_color="#f97316")
                    audio = r.listen(source, timeout=5, phrase_time_limit=10)
                    self.label_status.configure(text="PROCESSING...", text_color="#ef4444")
                    
                    command = r.recognize_google(audio)
                    self.update_transcript(command)
                    self.process_command(command)
                    
                except sr.WaitTimeoutError:
                    continue
                except Exception as e:
                    print(f"Error: {e}")
                    self.label_status.configure(text="AWAITING COMMAND", text_color="#f97316")

    def process_command(self, command):
        cmd = command.lower()
        
        if "youtube" in cmd:
            query = cmd.replace("search", "").replace("open", "").replace("youtube", "").replace("for", "").strip()
            if query:
                speak(f"Searching YouTube for {query}, Madan.")
                webbrowser.open(f"https://www.youtube.com/results?search_query={query}")
            else:
                speak("Opening YouTube, Madan.")
                webbrowser.open("https://www.youtube.com")
        
        elif "google" in cmd:
            query = cmd.replace("search", "").replace("open", "").replace("google", "").replace("for", "").strip()
            if query:
                speak(f"Searching Google for {query}, Madan.")
                webbrowser.open(f"https://www.google.com/search?q={query}")
            else:
                speak("Opening Google, Madan.")
                webbrowser.open("https://www.google.com")

        elif "whatsapp" in cmd:
            speak("Opening WhatsApp Web, Madan.")
            webbrowser.open("https://web.whatsapp.com")

        elif "gmail" in cmd:
            speak("Accessing your Gmail, Madan.")
            webbrowser.open("https://mail.google.com")

        else:
            # Gemini AI Response with 2026 context
            try:
                prompt = f"""
                User: {command}
                Context: You are JARVIS/FRIDAY v6.0 (2026 Edition). 
                Current Date: February 21, 2026.
                User Name: Madan.
                Instructions: Respond strictly in the language used (English or Telugu). 
                Use your knowledge of 2026 events. Keep it concise, loyal, and efficient.
                """
                response = model.generate_content(prompt)
                reply = response.text
                self.update_response(reply)
                speak(reply)
            except Exception as e:
                speak("I encountered a neural link error, Madan.")

if __name__ == "__main__":
    app = JarvisHUD()
    app.mainloop()
